### ISSUE TITLE
SSR Behavior

### ISSUE THREAD

**Issue by kaheglar** [cristianbote/goober#13]
Firstly nice lib, a minimal css-in-js solution is sorely needed!

From what I can see and there is a single global style cache and the extractCss() function will extract ALL cached styles? Is it then the case that, when rendering different content per page, ALL styles are embedded on each page (eventually)?

**Comment by cristianbote** [comment]
Hey Stephen, thanks for dropping by! 😃 

TL;DR; it's the same behaviour as using a _big_ `style` tag.

So, by doing SSR, you're actually sending down the styles already _collected_. On the client side, `goober` will pick-up the already sent style tag, and continue to append styles for component that are just rendered. And by `just rendered` that means, page navigation, component updates, and so on.

My main css-in-js experience was using emotion and styled-components, and that's how they work as well as far as I can debug them.

But I feel like, maybe you're concerned about something else? Is there a specific behaviour that you are concerned?

**Comment by kaheglar** [comment]
Hi Cristian,

OK I'll try to clarify. Imagine we have a server app that renders 3 pages (based on url) specified below:

PageA = ComponentA + StyleA
PageB = ComponentB + StyleB
PageC = ComponentC + StyleC

We start the server and the first four request come in as follows:

1. User1 requests PageA and receives "ComponentA + StyleA".
2. User2 requests PageB and receives "ComponentB + StyleA + Style B".
3. User3 requests PageC and receives "ComponentC + StyleA + StyleB + StyleC".
4. User1 requests PageA and receives "ComponentA + StyleA + StyleB + StyleC".

This happens because each of the single global sheet cache. This behaviour doesn't present in the browser as it's always single user context, where the server is multi user and each request is a unique context. Does this make sense?

**Comment by cristianbote** [comment]
Oh, now I get it 👍 .
You are correct! That's a mishap from my side 💯 . The proper behaviour would be, when `extractCss` would be called, the cache should be emptied, after the style tag is passed. I'll get to it in a moment.

Good call Stephen!

**Comment by kaheglar** [comment]
I guess this would work when the render call is synchronous and blocking, but what if it's not, e.g. ReactDOMServer.renderToNodeStream?

**Comment by cristianbote** [comment]
Hmm, now I know what you were _actually_ talking about 😃. That would change the API a bit, but not too much. Maybe it's gonna increase the size a bit, but I'll have to find some other stuff to adjust.

Let me think about it and do some testing. I'll get back with a _async_ API.

**Comment by cristianbote** [comment]
Hey @kaheglar, so I've done some research on how this could/should be implemented.

TL;DR; For now, because of the size increase, I'm tempted to not include it.

I have a rundown on how it could work and I was able to implemented it. So, given this server example:

```js
const http = require("http");
const { createElement } = require("react");
const { renderToNodeStream } = require("react-dom/server");

const Row = props => {
  return createElement(
    "div",
    null,
    props.index
  );
};

const App = props => {
  return createElement(
    "div",
    null,
    Array.from(Array(500)).map((_, i) =>
      createElement(Row, { key: i, index: i })
    )
  );
};

//create a server object:
http
  .createServer(function(req, res) {
    res.write(`
      <!DOCTYPE html>
      <html>
        <head></head>
        <body>
          <div id="root">`);

    const stream = renderToNodeStream(
      createElement(App)
    );

    stream.pipe(res, { end: "false" });
    stream.on("end", () => res.end(`</div></body></html>`));
  })
  .listen(8080); //the server object listens on port 8080
```

So to manage to write on the stream we need to create our own stream and pipe the `res` to it. And for each `add()` call from `goober` push it to the stream.

So this is how it would look like:
```js
const http = require("http");
const { Readable } = require("stream");
const { styled, setPragma, setOnAdd, extractCss } = require("./goober");
const { createElement } = require("react");
const { renderToNodeStream } = require("react-dom/server");

// Se the pragma for vnode creation
setPragma(createElement);

const StyledButton = styled("button")`
  border: 0;
  border-radius: 4px;
  background: dodgerblue;
  color: white;
  font-size: 1em;
`;

const StyledRow = styled("div")`
  background: dodgerblue;
  color: white;
  float: left;
  display: inline-block;

  &:nth-child(2n + 2) {
    background: darkblue;
  }
`;

const Row = props => {
  return createElement(
    StyledRow,
    null,
    createElement(StyledButton, null, props.index)
  );
};

const App = props => {
  return createElement(
    "div",
    null,
    Array.from(Array(500)).map((_, i) =>
      createElement(Row, { key: i, index: i })
    )
  );
};

//create a server object:
http
  .createServer(function(req, res) {

    // TODO: First change
    // This should be just a simple _flush_, so that's another function
    const css = extractCss();

    res.write(`
      <!DOCTYPE html>
      <html>
        <head></head>
        <body>
          <div id="root">`);

    let stream = renderToNodeStream(
      createElement(App)
    );

    // Our readable stream to push to
    const src = new Readable({
      read() {
        return "";
      }
    });

    // TODO: Second change.
    // This will be called for each style add
    setOnAdd(css => src.push(`<style data-goober-stream>${css}</style>`));

    src.pipe(
      res,
      { end: "false" }
    );

    stream.pipe(
      res,
      { end: "false" }
    );

    stream.on("end", () => res.end(`</div></body></html>`));
  })
  .listen(8080); //the server object listens on port 8080
```

All in all, this would add `+20B` 😞. But all is not lost. @michael-klein is working on a custom `target` implementation that could give us more space and maybe another pattern to implement this.

I'm gonna close it, just so I can mark it. But, no worries, once `goober` would add support for this I'm sure gonna add a shoutout here.

**Comment by ksbrooksjr** [comment]
Hey Cristian, thanks for your work on Goober. Is there any update on streaming SSR support? Without streaming you have to block the event loop while synchronously rendering using renderToString.
